diff --git a/.vscode/settings.json b/.vscode/settings.json
index 164e3ce..93ee22d 100644
--- a/.vscode/settings.json
+++ b/.vscode/settings.json
@@ -43,6 +43,30 @@
         "unordered_map": "cpp",
         "vector": "cpp",
         "xhash": "cpp",
-        "string": "cpp"
+        "string": "cpp",
+        "array": "cpp",
+        "atomic": "cpp",
+        "bit": "cpp",
+        "*.tcc": "cpp",
+        "cctype": "cpp",
+        "clocale": "cpp",
+        "codecvt": "cpp",
+        "condition_variable": "cpp",
+        "cstdarg": "cpp",
+        "ctime": "cpp",
+        "cwctype": "cpp",
+        "deque": "cpp",
+        "map": "cpp",
+        "algorithm": "cpp",
+        "iterator": "cpp",
+        "memory_resource": "cpp",
+        "numeric": "cpp",
+        "optional": "cpp",
+        "random": "cpp",
+        "string_view": "cpp",
+        "fstream": "cpp",
+        "mutex": "cpp",
+        "sstream": "cpp",
+        "cinttypes": "cpp"
     }
 }
\ No newline at end of file
diff --git a/app.cc b/app.cc
index 8ef220e..efa962b 100755
--- a/app.cc
+++ b/app.cc
@@ -73,6 +73,7 @@ App::App() : udsclient_(log_),
     log_.SetLogLvl(LOG_DEBUG);
     log_.SetPriv(std::string (LOG_DIR_PATH), MyLogFile::LogType::Daily, std::string("adab.log"), std::string (""));
     log_.LogWrite(LOG_INFO, std::string ("ADAB|start"));
+    log_.LogWrite(LOG_INFO, std::string ("ADAB|VERSION : 0.7"));
 
     if(InitGpioC2MReset() == false) {
         log_.LogWrite(LOG_ERR, std::string ("ADAB|InitGpioC2MReset fail"));
@@ -528,4 +529,4 @@ void App::WriteAdabErrStatus(int err) {
 	myfile.open(ADABSTATUS_PATH, std::ios_base::out);
     myfile << sstream.str();
 	myfile.close();
-}
\ No newline at end of file
+}
diff --git a/buffer.cc b/buffer.cc
index 1f5e9f3..c15a082 100755
--- a/buffer.cc
+++ b/buffer.cc
@@ -6,10 +6,12 @@
 
 #include <sys/time.h>
 
+#define MAX_QUEUE_SIZE (1ULL << (4 * 8 - 1))
+
 BufferBase::BufferBase(MyLogFile &log) : log_(log) {
     q_head_ = 0;
     q_tail_ = 0;
-    q_count_ = 2048;
+    q_count_ = 2048*8;
 }
 
 BufferBase::~BufferBase() {
@@ -25,23 +27,76 @@ void BufferBase::Purge(void) {
     q_mutex_.lock();
     q_head_ = 0;
     q_tail_ = 0;
-    q_count_ = 2048;
+    q_count_ = 2048*8;
     q_mutex_.unlock();
 }
 
+void BufferBase::IncreaseQueueSize() {
+    size_t new_size = q_count_ * 2;  // 큐 크기를 두 배로 증가
+    if (new_size > MAX_QUEUE_SIZE) {
+        new_size = MAX_QUEUE_SIZE;  // 최대 큐 크기를 초과하지 않도록 제한
+    }
+
+    uint8_t* new_buf = (uint8_t*)malloc(new_size * buf_size_);
+    if (!new_buf) {
+        std::cout << "Memory allocation failed during queue resizing." << std::endl;
+        return;
+    }
+
+    // 기존 데이터를 새 버퍼로 복사
+    for (size_t i = 0; i < q_count_; ++i) {
+        memcpy(new_buf + i * buf_size_, q_buf_ + ((q_head_ + i) % q_count_) * buf_size_, buf_size_);
+    }
+
+    free(q_buf_);  // 기존 버퍼 해제
+    q_buf_ = new_buf;
+    q_count_ = new_size;
+    q_head_ = 0;
+    q_tail_ = q_count_ / 2;
+}
+
+#if 1
 bool BufferBase::Push(const void *ptr) {
     size_t new_tail;
     q_mutex_.lock();
     new_tail = (q_tail_ + 1) % q_count_;
+
+#if 0
     if(new_tail == q_head_) {
         q_mutex_.unlock();
         return false;
     }
+#endif
+
+    if (new_tail == q_head_) {
+        std::string cal_msg;
+        char buffer[256];
+
+        if (q_count_ >= MAX_QUEUE_SIZE) {
+            // 큐의 크기가 최대치를 넘으면 false를 반환
+            q_mutex_.unlock();
+            //std::cout << "Queue size limit reached. Unable to increase size." << std::endl;
+            sprintf(buffer, "q_tail:%ld, new_tail:%ld, q_head:%ld, q_count:%ld, q_max:%lld", q_tail_, new_tail, q_head_, q_count_, MAX_QUEUE_SIZE);
+            cal_msg = "ADAB|queue full, ";
+            cal_msg += buffer;
+            log_.LogWrite(LOG_ERR, cal_msg);
+            return false;
+        }
+
+        IncreaseQueueSize();  // 큐 크기를 증가시키는 함수
+        //std::cout << "push new_tail : "<< new_tail << ", q_head :"<< q_head_ << ", q_count : "<< q_count_ <<std::endl;
+        sprintf(buffer, "q_tail:%ld, new_tail:%ld, q_head:%ld, q_count:%ld, q_max:%lld", q_tail_, new_tail, q_head_, q_count_, MAX_QUEUE_SIZE);
+        cal_msg = "ADAB|queue increase size, ";
+        cal_msg += buffer;
+        log_.LogWrite(LOG_WARNING, cal_msg);
+    }
 
     memcpy(q_buf_ + q_tail_*buf_size_, ptr, buf_size_);
     q_tail_ = new_tail;
     q_mutex_.unlock();
-    cv_.notify_all();
+
+    //cv_.notify_all();
+    cv_.notify_one();
 
     return true;
 }
@@ -69,4 +124,73 @@ bool BufferBase::Pop(void *ptr, uint64_t *timestamp) {
     q_mutex_.unlock();
 
     return true;
-}
\ No newline at end of file
+}
+
+#else
+
+bool BufferBase::Push(const void *ptr) {
+    size_t new_tail;
+    q_mutex_.lock();
+    new_tail = (q_tail_ + 1) % q_count_;
+
+    // 큐가 가득 차면 크기를 증가시킴
+    if (new_tail == q_head_) {
+        std::string cal_msg;
+        char buffer[256];
+
+        if (q_count_ >= MAX_QUEUE_SIZE) {
+            // 큐의 크기가 최대치를 넘으면 false를 반환
+            q_mutex_.unlock();
+            //std::cout << "Queue size limit reached. Unable to increase size." << std::endl;
+            sprintf(buffer, "q_tail:%ld, new_tail:%ld, q_head:%ld, q_count:%ld, q_max:%lld", q_tail_, new_tail, q_head_, q_count_, MAX_QUEUE_SIZE);
+            cal_msg = "ADAB|queue full, ";
+            cal_msg += buffer;
+            log_.LogWrite(LOG_ERR, cal_msg);
+            return false;
+        }
+
+        IncreaseQueueSize();  // 큐 크기를 증가시키는 함수
+        //std::cout << "push new_tail : "<< new_tail << ", q_head :"<< q_head_ << ", q_count : "<< q_count_ <<std::endl;
+        sprintf(buffer, "q_tail:%ld, new_tail:%ld, q_head:%ld, q_count:%ld, q_max:%lld", q_tail_, new_tail, q_head_, q_count_, MAX_QUEUE_SIZE);
+        cal_msg = "ADAB|queue increase size, ";
+        cal_msg += buffer;
+        log_.LogWrite(LOG_WARNING, cal_msg);
+    }
+
+    // 데이터 복사
+    memcpy(q_buf_ + q_tail_ * buf_size_, ptr, buf_size_);
+
+    // tail 포인터 업데이트
+    q_tail_ = new_tail;
+    q_mutex_.unlock();
+    cv_.notify_one();
+
+    return true;
+}
+
+bool BufferBase::Pop(void *ptr, uint64_t *timestamp) {
+    std::unique_lock<std::mutex> lk(cv_m_);
+    if (q_head_ == q_tail_) {
+        // 대기 시간 설정
+        auto now = std::chrono::system_clock::now();
+        if (!cv_.wait_until(lk, now + std::chrono::milliseconds(10), [this] { return q_head_ != q_tail_; })) {
+            return false;
+        }
+
+        if (*timestamp == 0) {
+            struct timeval time;
+            gettimeofday(&time, NULL);
+            *timestamp = (uint64_t)time.tv_sec * 1000000 + (uint64_t)time.tv_usec;
+        }
+    }
+
+    // 큐에서 데이터 복사
+    q_mutex_.lock();
+    memcpy(ptr, q_buf_ + q_head_ * buf_size_, buf_size_);
+    q_head_ = (q_head_ + 1) % q_count_;
+    q_mutex_.unlock();
+    //std::cout << "pop q_head : "<< q_head_ << std::endl;
+    
+    return true;
+}
+#endif
\ No newline at end of file
diff --git a/buffer.h b/buffer.h
index 4f3137c..5aaf339 100755
--- a/buffer.h
+++ b/buffer.h
@@ -8,6 +8,7 @@
 
 #include <stddef.h>
 #include <stdint.h>
+#include <atomic>
 
 struct __attribute__ ((packed)) buffer_item {
     uint32_t tick;
@@ -39,6 +40,7 @@ public:
     bool Push(const void *ptr);
     bool Pop(void *ptr, uint64_t *timestamp);
     void Purge(void);
+    void IncreaseQueueSize();
     virtual int GetType(void) = 0;
     virtual size_t GetBufSize(void) = 0;
     virtual uint32_t GetTick(uint8_t* buf) = 0;
@@ -55,6 +57,8 @@ private:
     size_t q_count_;
     size_t q_head_;
     size_t q_tail_;
+    //std::atomic<size_t> q_tail_;
+    //std::atomic<size_t> q_head_;
     std::mutex q_mutex_;
     std::condition_variable cv_;
     std::mutex cv_m_;
